File: webapp/src/util/instrument.js
import { toMoney } from "@/util/utils";
import { SideString } from "@/util/option";

export function instrumentCardTitle(instrument, side, locale) {
  let i = instrument;
  if (!i) {
    return;
  }
  let sideString = SideString(side, instrument.optionType, locale);
  return `$${toMoney(i.strikePrice)} ${sideString} ${i.baseCurrency}`;
}

export function parseInstrumentName(name) {
  let idxDeliveryType = 1;
  let idxQuoteCurrency = 2;
  let idxBaseCurrency = 3;
  let idxDay = 4;
  let idxMonth = 5;
  let idxYear = 6;
  let idxStrikePrice = 7;
  let idxOptionType = 8;

  let instrument = {};
  let monthMap = {
    JAN: "01",
    FEB: "02",
    MAR: "03",
    APR: "04",
    MAY: "05",
    JUN: "06",
    JUL: "07",
    AUG: "08",
    SEP: "09",
    OCT: "10",
    NOV: "11",
    DEC: "12",
  };
  let reg =
    /^([CP])-(pUSD|USDT)-(BTC|XIN|ETH)-([\d]{1,2})(JAN|FEB|MAR|APR|MAY|JUN|JUL|AUG|SEP|OCT|NOV|DEC)([\d]{2})-([\d]+)-([CP])$/;
  let match = name.match(reg);
  let day = match[idxDay];
  if (day.length === 1) {
    day = "0" + day;
  }
  let timeString = `20${match[idxYear]}-${match[idxMonth]}-${day}T16:00:00+08:00`;
  let expirationDate = new Date(timeString);
  let strikePrice = match[idxStrikePrice];
  let deliveryType = "";
  if (match[idxDeliveryType] === "P") {
    deliveryType = "PHYSICAL";
  }
  if (match[idxDeliveryType] === "C") {
    deliveryType = "CASH";
  }
  let optionType = "";
  if (match[idxOptionType] === "P") {
    optionType = "PUT";
  }
  if (match[idxOptionType] === "C") {
    optionType = "CALL";
  }
  if (match.length === 9) {
    instrument = {
      deliveryType: deliveryType,
      quoteCurrency: match[idxQuoteCurrency],
      baseCurrency: match[idxBaseCurrency],
      expirationDate: expirationDate,
      expirationTimestamp: expirationDate.getTime(),
      strikePrice: strikePrice,
      optionType: optionType,
    };
  }
  return instrument;
}

// The deribit rule, tomorrow is the day after tomorrow, the day after tomorrow +7 the day after tomorrow +14,
// (the day after tomorrow +14) the last Friday of the next month, (the day after tomorrow +14)
export function getBidExpiryDates() {
  let dates = [];
  let firstDay = new Date();
  firstDay.setDate(firstDay.getDate() + 1);
  dates.push(firstDay.Format("yyyy/MM/dd"));
  let secondDay = new Date(firstDay);
  secondDay.setDate(secondDay.getDate() + 1);
  let lastWeeksDay = new Date();
  let weeksDay = Array.from({ length: 3 }, (item, index) => {
    if (index === 0) {
      return secondDay.Format("yyyy/MM/dd");
    } else {
      let date = new Date(secondDay);
      date.setDate(secondDay.getDate() + 7 * index);
      lastWeeksDay = date;
      return date.Format("yyyy/MM/dd");
    }
  });
  let result = dates.concat(weeksDay);

  let diffList = [1, 2, 4, 7, 10];
  let year = lastWeeksDay.getFullYear();
  let isYearIncr = false;
  for (let i in diffList) {
    let month = lastWeeksDay.getMonth() + diffList[i];
    if (month > 11 && !isYearIncr) {
      year++;
      isYearIncr = true;
    }
    let lastFriday = lastFridayOfMonth(year, month);
    result = result.concat(lastFriday.Format("yyyy/MM/dd"));
  }

  return result;
}

export function toExpiryDate(yyMMdd) {
  let strings = yyMMdd.split("/");
  if (strings.length > 1) {
    let year = Number(strings[0]);
    let monthIdx = Number(strings[1]) - 1;
    let day = Number(strings[2]);
    let date = new Date(year, monthIdx, day);
    date.setHours(16, 0, 0, 0);
    return date;
  } else {
    return null;
  }
}

export function leftSettlementPeriod(expirationDate) {
  if (expirationDate) {
    let now = new Date();
    let nowTs = now.getTime();
    let expiry = new Date(expirationDate);
    let expiryTs = new Date(expirationDate).getTime();
    let differenceInTime = expiryTs - nowTs;
    let day = parseInt(differenceInTime / (1000 * 3600 * 24));
    if (day === 0) {
      let hour = parseInt(differenceInTime / (1000 * 3600));
      return `${hour} hours`;
    }
    return `${day + 1} days`;
  } else {
    return "0 days";
  }
}

function lastFridayOfMonth(year, month) {
  "use strict";
  var lastDay = new Date(year, month + 1, 0);
  if (lastDay.getDay() < 6) {
    lastDay.setDate(lastDay.getDate() - 7);
  }
  lastDay.setDate(lastDay.getDate() - (lastDay.getDay() - 6));
  return lastDay;
}

File: core/instrument.go
package core

import (
	"errors"
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"time"
)

var (
	WrongInstrumentFormatErr = errors.New("wrong instrument format")
)

const (
	OptionTypePUT        = "PUT"
	OptionTypeCALL       = "CALL"
	DeliveryTypeCash     = "CASH"
	DeliveryTypePhysical = "PHYSICAL"

	// 8：00 ~ 16：00
	exercisePeriod = 8 * 60 * 60
	//16:00 ~ 16:30
	expiryToSettlementPeriod = 30 * 60
)

const (
	idxDeliveryType = iota + 1
	idxQuoteCurrency
	idxBaseCurrency
	idxDay
	idxMonth
	idxYear
	idxStrikePrice
	idxOptionType
)

var (
	monthEnMap = map[string]string{
		"01": "JAN", "02": "FEB", "03": "MAR", "04": "APR", "05": "MAY", "06": "JUN",
		"07": "JUL", "08": "AUG", "09": "SEP", "10": "OCT", "11": "NOV", "12": "DEC",
	}
)

// To determine whether it can be exercised
func Exercisable(exercisedTs, expiryTs int64) bool {
	settlementTs := expiryTs + expiryToSettlementPeriod
	return exercisedTs < settlementTs && exercisedTs > expiryTs-exercisePeriod
}

func ExercisableContract(exercisedTs, expiryTs int64) bool {
	return exercisedTs > expiryTs-exercisePeriod
}

// To determine whether an instrument is expired
func Expired(createAtTs, expiryTs int64) bool {
	return createAtTs >= expiryTs
}

func ExpiredPosition(createAtTs, expiryTs int64) bool {
	return createAtTs >= expiryTs+expiryToSettlementPeriod
}

type OptionInfo struct {
	DeliveryType        string
	QuoteCurrency       string
	BaseCurrency        string
	ExpirationDate      time.Time
	ExpirationTimestamp int64
	StrikePrice         int64
	OptionType          string
}

// ParseInstrument eg. format C-pUSD-BTC-24SEP21-80000-P
func ParseInstrument(i string) (info OptionInfo, err error) {
	monthMap := map[string]string{"JAN": "01", "FEB": "02", "MAR": "03", "APR": "04", "MAY": "05", "JUN": "06", "JUL": "07", "AUG": "08", "SEP": "09", "OCT": "10", "NOV": "11", "DEC": "12"}
	reg := `^(C|P)-(pUSD|USDT|BTC)-(BTC|XIN|ETH)-([\d]{1,2})(JAN|FEB|MAR|APR|MAY|JUN|JUL|AUG|SEP|OCT|NOV|DEC)([\d]{2})-([\d]+)-(C|P)$`
	instrumentReg, err := regexp.Compile(reg)
	if err != nil {
		return info, err
	}
	params := instrumentReg.FindStringSubmatch(i)

	var deliveryType string
	if params[idxDeliveryType] == "P" {
		deliveryType = DeliveryTypePhysical
	} else if params[idxDeliveryType] == "C" {
		deliveryType = DeliveryTypeCash
	} else {
		return info, fmt.Errorf("instrument deliveryType error: %s", i)
	}

	var day = params[idxDay]
	if len(day) == 1 {
		day = "0" + day
	}
	timeString := fmt.Sprintf("20%s-%s-%sT16:00:00+08:00", params[idxYear], monthMap[params[idxMonth]], day)
	expirationDate, err := time.Parse(time.RFC3339, timeString)
	if err != nil {
		return info, err
	}
	strikePrice, err := strconv.Atoi(params[idxStrikePrice])
	if err != nil {
		return info, err
	}
	var optionType string
	if params[idxOptionType] == "P" {
		optionType = OptionTypePUT
	} else if params[idxOptionType] == "C" {
		optionType = OptionTypeCALL
	} else {
		return info, fmt.Errorf("instrument optionType error: %s", i)
	}
	if len(params) == 9 {
		info := OptionInfo{
			DeliveryType:        deliveryType,
			QuoteCurrency:       params[idxQuoteCurrency],
			BaseCurrency:        params[idxBaseCurrency],
			ExpirationDate:      expirationDate,
			ExpirationTimestamp: expirationDate.UTC().Unix(),
			StrikePrice:         int64(strikePrice),
			OptionType:          optionType,
		}
		return info, nil
	} else {
		return info, fmt.Errorf("instrument format error: %s", i)
	}
}

func InstrumentDate(time time.Time) string {
	format := time.Format("06-01-02")
	split := strings.Split(format, "-")
	monthMap := map[string]string{"01": "JAN", "02": "FEB", "03": "MAR", "04": "APR", "05": "MAY", "06": "JUN", "07": "JUL", "08": "AUG", "09": "SEP", "10": "OCT", "11": "NOV", "12": "DEC"}
	day := split[2]
	if strings.HasPrefix(day, "0") {
		day = day[1:]
	}
	return fmt.Sprintf("%s%s%s", day, monthMap[split[1]], split[0])
}

func InstrumentExpiryTimestamp(t time.Time) int64 {
	date := time.Date(t.Year(), t.Month(), t.Day(), 8, 0, 0, 0, time.UTC)
	return date.Unix()
}

func InstrumentExpiryTs(year, month, day int) int64 {
	return InstrumentTime(year, month, day).Unix()
}

func InstrumentTime(year, month, day int) time.Time {
	return time.Date(year, time.Month(month), day, 8, 0, 0, 0, time.UTC)
}

func ParseOptionType(instrument string) string {
	var optionType = OptionTypePUT
	if strings.HasSuffix(instrument, "-C") {
		optionType = OptionTypeCALL
	}
	return optionType
}

// ToOdInstrument eg. format C-pUSD-BTC-24SEP21-80000-P
func ToOdInstrument(deliveryType, quoteCurrency, baseCurrency, strikePrice, optionType string, expiryDate time.Time) string {
	var (
		dt = "C"
		qc = "pUSD"
		bc = "BTC"
		ep = ""
		sp = strikePrice
		ot = "P"
	)
	if deliveryType == DeliveryTypePhysical {
		dt = "P"
	}
	if quoteCurrency != "" {
		qc = quoteCurrency
	}
	if baseCurrency != "" {
		bc = baseCurrency
	}
	if optionType == OptionTypeCALL {
		ot = "C"
	}
	//expiryDate
	format := expiryDate.Format("020106")
	day := format[:2]
	if strings.HasPrefix(day, "0") {
		day = day[1:]
	}
	ep = fmt.Sprintf("%s%s%s", day, monthEnMap[format[2:4]], format[4:])
	return fmt.Sprintf("%s-%s-%s-%s-%s-%s", dt, qc, bc, ep, sp, ot)
}

// ToDbInstrument eg. format BTC-24SEP21-80000-P
func ToDbInstrument(baseCurrency, strikePrice, optionType string, expiryDate time.Time) string {
	var (
		bc = "BTC"
		ep = ""
		sp = strikePrice
		ot = "P"
	)
	if baseCurrency != "" {
		bc = baseCurrency
	}
	if optionType == OptionTypeCALL {
		ot = "C"
	}
	//expiryDate
	format := expiryDate.Format("020106")
	day := format[:2]
	if strings.HasPrefix(day, "0") {
		day = day[1:]
	}
	ep = fmt.Sprintf("%s%s%s", day, monthEnMap[format[2:4]], format[4:])
	return fmt.Sprintf("%s-%s-%s-%s", bc, ep, sp, ot)
}
